polar-core/src/bindings.rs

```diff
@@ -179,35 +194,36 @@ impl BindingManager {
     ///
     /// If a binding between two variables is made, and one is bound and the other unbound, the
     /// unbound variable will take the value of the bound one.
-    pub fn bind(&mut self, var: &Symbol, val: Term) -> PolarResult<()> {
+    pub fn bind(&mut self, var: &Symbol, val: Term) -> PolarResult<Option<Goal>> {
+        let mut goal = None;
         if let Ok(symbol) = val.value().as_symbol() {
-            self.bind_variables(var, symbol)?;
-        } else if let BindingManagerVariableState::Partial(p) = self._variable_state(var) {
-            if let Some(grounded) = p.ground(var.clone(), val.clone()) {
-                self.add_binding(var, val.clone());
-                self.constrain(&grounded)?;
-            } else {
-                return Err(RuntimeError::IncompatibleBindings {
-                    msg: "Grounding failed".into(),
-                }
-                .into());
-            }
+            goal = self.bind_variables(var, symbol)?;
         } else {
-            if let BindingManagerVariableState::Bound(_) = self._variable_state(var) {
-                return Err(RuntimeError::IncompatibleBindings {
-                    msg: format!("Cannot rebind {:?}", var),
+            match self._variable_state(var) {
+                BindingManagerVariableState::Partial(p) => {
+                    let p = p.clone();
+                    let val = val.clone();
+                    goal = Some(self.partial_bind(p, var, val)?)
                 }
-                .into());
-            }
 
-            self.add_binding(var, val.clone());
+                BindingManagerVariableState::Bound(_) => {
+                    return Err(RuntimeError::IncompatibleBindings {
+                        msg: format!("Cannot rebind {:?}", var),
```

Comment:
This is a good example of where runtime error is the correct behaviour -- the VM shouldn't be reaching this state where it attempts to rebind a value.

Reply:
Oh yeah, I was thinking of `OperationalError` which we have to represent these internal errors.